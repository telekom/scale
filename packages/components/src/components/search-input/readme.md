# scale-search



<!-- Auto Generated Below -->


## Properties

| Property                 | Attribute                 | Description                                                                                                      | Type      | Default                     |
| ------------------------ | ------------------------- | ---------------------------------------------------------------------------------------------------------------- | --------- | --------------------------- |
| `ariaDetailedId`         | `aria-detailed-id`        | (optional) id or space separated list of ids of elements that provide or link to additional related information. | `string`  | `undefined`                 |
| `disabled`               | `disabled`                | (optional) Input disabled                                                                                        | `boolean` | `undefined`                 |
| `dismissable`            | `dismissable`             |                                                                                                                  | `boolean` | `undefined`                 |
| `experimentalControlled` | `experimental-controlled` | (optional)) Makes type `input` behave as a controlled component in React                                         | `boolean` | `false`                     |
| `hideLabelVisually`      | `hide-label-visually`     | (optional) to avoid displaying the label                                                                         | `boolean` | `false`                     |
| `inputAutocomplete`      | `input-autocomplete`      | (optional) custom value for autocomplete HTML attribute                                                          | `string`  | `undefined`                 |
| `inputAutofocus`         | `input-autofocus`         | (optional) the input should automatically get focus when the page loads.                                         | `boolean` | `undefined`                 |
| `inputId`                | `input-id`                | (optional) Input id                                                                                              | `string`  | `undefined`                 |
| `invalid`                | `invalid`                 | (optional) Input status                                                                                          | `boolean` | `false`                     |
| `maxLength`              | `max-length`              | (optional) Input text string max length                                                                          | `number`  | `undefined`                 |
| `minLength`              | `min-length`              | (optional) Input text string min length                                                                          | `number`  | `undefined`                 |
| `name`                   | `name`                    | (optional) Input name                                                                                            | `string`  | `'Search'`                  |
| `placeholder`            | `placeholder`             | (optional) Input placeHolder                                                                                     | `string`  | `'Search'`                  |
| `required`               | `required`                | (optional) Input required                                                                                        | `boolean` | `undefined`                 |
| `styles`                 | `styles`                  | (optional) Injected CSS styles                                                                                   | `string`  | `undefined`                 |
| `transparent`            | `transparent`             | (optional) input background transparent                                                                          | `boolean` | `undefined`                 |
| `value`                  | `value`                   | (optional) Input value                                                                                           | `string`  | `''`                        |
| `variant`                | `variant`                 |                                                                                                                  | `string`  | `'onHover' \|\| 'alwaysOn'` |


## Events

| Event                          | Description                             | Type                                  |
| ------------------------------ | --------------------------------------- | ------------------------------------- |
| `scale-blur`                   | Emitted when the input loses focus.     | `CustomEvent<void>`                   |
| `scale-change`                 | Emitted when the value has changed.     | `CustomEvent<InputChangeEventDetail>` |
| `scale-focus`                  | Emitted when the input has focus.       | `CustomEvent<void>`                   |
| `scale-focus-out`              | Emitted when the input has focus.       | `CustomEvent<void>`                   |
| `scale-input`                  | Emitted when a keyboard input occurred. | `CustomEvent<KeyboardEvent>`          |
| `scale-interactive-icon-click` | Emitted on interactive icon click       | `CustomEvent<KeyboardEvent>`          |
| `scale-keydown`                | Emitted on keydown.                     | `CustomEvent<KeyboardEvent>`          |


## Shadow Parts

| Part                  | Description |
| --------------------- | ----------- |
| `"clear-icon"`        |             |
| `"clear-icon-button"` |             |
| `"interactive-icon"`  |             |


## Dependencies

### Depends on

- [scale-icon-button](../icon-button)
- [scale-icon-action-close](../icons/action-close)

### Graph
```mermaid
graph TD;
  scale-search-input --> scale-icon-button
  scale-search-input --> scale-icon-action-close
  style scale-search-input fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
