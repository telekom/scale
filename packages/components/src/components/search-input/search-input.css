/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

:host {
  /* base */
  --width: 100%;
  --min-width: 144px;
  --height: 44px;
  --border: 1px solid var(--telekom-color-ui-border-hovered);
  --border-radius: var(--telekom-spacing-unit-x2);
  --margin: 0;
  /* focus */
  --focus-box-shadow: transparent;
  --focus-outline: var(--telekom-spacing-unit-x05) solid
    var(--telekom-color-functional-focus);
  --focus-outline-offset: var(--telekom-spacing-unit-x025);
  --focus-prefix-icon-color: var(--telekom-color-text-and-icon-standard);
  /* hover */
  --hover-background-color: var(--telekom-color-ui-state-fill-hovered);
  --hover-border-color: var(--telekom-color-ui-border-hovered);
  /* active */
  --active-background-color: var(--telekom-color-ui-state-fill-hovered);
  /* disabled */
  --disabled-slot_input-border-color: var(--telekom-color-ui-border-disabled);
  --disabled-slot_input-color: var(--telekom-color-text-and-icon-disabled);

  /* input */
  --input-width: 100%;
  --input-height: 44px;
  --input-padding-left: var(--telekom-spacing-unit-x2);
  --input-margin: 0;
  --input-font-family: inherit;
  --input-font-size: var(--telekom-typography-font-size-body);
  --input-background-color: var(--telekom-color-ui-state-fill-standard);

  /* interactive-icon */
  --interactive-icon-color: var(--telekom-color-text-and-icon-additional);

  /* clear-icon-button */
  --clear-icon-button-margin-left: var(--telekom-spacing-unit-x2);
  --clear-icon-button-margin-right: var(--telekom-spacing-unit-x2);

  /* clear-icon */
  --clear-icon-color: var(--telekom-color-text-and-icon-standard);

  /* suffix-icon */
  --suffix-icon-margin-left: var(--telekom-spacing-unit-x3);
  --suffix-icon-margin-right: var(--telekom-spacing-unit-x2);
  --suffix-icon-color: var(--telekom-color-text-and-icon-additional);

  /* prefix-icon */
  --prefix-icon-margin-left: var(--telekom-spacing-unit-x3);
  --prefix-icon-color: var(--telekom-color-text-and-icon-additional);

  /* placeholder */
  --placeholder-color: var(--telekom-color-text-and-icon-additional);
}

[part~='base'] {
  width: var(--width);
  height: var(--height);
  border: var(--border);
  border-radius: var(--border-radius);
  margin: var(--margin);
  display: flex;
  outline: none;
  min-width: var(--min-width);
}

[part~='base'][part~='focus'],
[part~='base']:focus {
  box-shadow: var(--focus-box-shadow);
  outline: var(--focus-outline);
  outline-offset: var(--focus-outline-offset);
  z-index: 40;
}

[part~='base']:hover {
  background-color: var(--hover-background-color);
  border-color: var(--hover-border-color);
}

[part~='base']:active {
  background-color: var(--active-background-color);
}

[part~='input'] {
  width: var(--input-width);
  height: var(--input-height);
  padding-left: var(--input-padding-left);
  margin: var(--input-margin);
  display: flex;
  outline: none;
  border: 0;
  box-sizing: border-box;
  font-family: var(--input-font-family);
  font-size: var(--input-font-size);
  background-color: var(--input-background-color);
}

[part~='interactive-icon'] {
  align-items: center;
  display: flex;
  color: var(--interactive-icon-color);
}

[part~='clear-icon-button'] {
  margin-left: var(--clear-icon-button-margin-left);
  margin-right: var(--clear-icon-button-margin-right);
  align-items: center;
  --background: transparent;
}

[part~='clear-icon'] {
  color: var(--clear-icon-color);
}

slot[name='prefix-icon']::slotted(*) {
  margin-left: var(--prefix-icon-margin-left);
  align-items: center;
  color: var(--prefix-icon-color);
}

[part~='base'][part~='focus'] slot[name='prefix-icon']::slotted(*) {
  color: var(--focus-prefix-icon-color);
}

slot[name='suffix-icon']::slotted(*) {
  margin-left: var(--suffix-icon-margin-left);
  margin-right: var(--suffix-icon-margin-right);
  align-items: center;
  color: var(--suffix-icon-color);
  --background: transparent;
  --background-hover: var(--telekom-color-ui-state-fill-hovered);
  --color: var(--suffix-icon-color);
}

[part~='base']:hover slot[name='suffix-icon']::slotted(*) {
  --background: transparent;
}

[part~='base'][part~='focus'] slot[name='suffix-icon']::slotted(*) {
  --color: var(--telekom-color-text-and-icon-standard);
}

[part~='base'][part~='disabled'] input,
[part~='base'][part~='disabled'] slot {
  cursor: not-allowed;
  border-color: var(--disabled-slot_input-border-color);
  background-color: none;
  color: var(--disabled-slot_input-color);
  background: transparent;
}

/* Hide default styling for type search */

input[type='search']::-webkit-search-decoration,
input[type='search']::-webkit-search-cancel-button,
input[type='search']::-webkit-search-results-button,
input[type='search']::-webkit-search-results-decoration {
  -webkit-appearance: none;
}

/*Hide Arrows From Input Number*/

/* Chrome, Safari, Edge, Opera */
input::-webkit-outer-spin-button,
input::-webkit-inner-spin-button {
  -webkit-appearance: none;
}

::placeholder {
  /* Chrome, Firefox, Opera, Safari 10.1+ */
  color: var(--placeholder-color);
  opacity: 1; /* Firefox */
}

:-ms-input-placeholder {
  /* Internet Explorer 10-11 */
  color: var(--placeholder-color);
}

::-ms-input-placeholder {
  /* Microsoft Edge */
  color: var(--placeholder-color);
}
