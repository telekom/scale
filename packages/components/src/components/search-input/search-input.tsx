/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import {
  Component,
  Prop,
  Element,
  Event,
  h,
  EventEmitter,
  Host,
  State,
} from '@stencil/core';
import classNames from 'classnames';
import { emitEvent, generateUniqueId } from '../../utils/utils';

interface InputChangeEventDetail {
  value: string | undefined | null;
}

@Component({
  tag: 'scale-search-input',
  styleUrl: './search-input.css',
  shadow: true,
})
export class SearchInput {
  @Element() hostElement: HTMLElement;
  /** (optional) Input name */
  @Prop() name?: string = 'Search';
  /** (optional) Input status */
  @Prop() invalid?: boolean = false;
  /** (optional) Input text string max length */
  @Prop() maxLength?: number;
  /** (optional) Input text string min length */
  @Prop() minLength?: number;
  /** (optional) Input placeHolder */
  @Prop() placeholder?: string = 'Search';
  /** (optional) Input disabled */
  @Prop() disabled?: boolean;
  /** (optional) Input required */
  @Prop() required?: boolean;
  /** (optional) Input value */
  @Prop({ mutable: true }) value?: string | null = '';
  /** (optional) Input id */
  @Prop() inputId?: string;
  /** (optional) input background transparent */
  @Prop() transparent?: boolean;
  /** (optional) the input should automatically get focus when the page loads. */
  @Prop() inputAutofocus?: boolean;
  /** (optional) custom value for autocomplete HTML attribute */
  @Prop() inputAutocomplete?: string;
  /** (optional) id or space separated list of ids of elements that provide or link to additional related information. */
  @Prop() ariaDetailedId?: string;
  /** (optional) to avoid displaying the label */
  @Prop() hideLabelVisually?: boolean = false;
  /** (optional) Injected CSS styles */
  @Prop() styles?: string;
  /** (optional)) Makes type `input` behave as a controlled component in React */
  @Prop() experimentalControlled?: boolean = false;
  /** Emitted when a keyboard input occurred. */
  @Event({ eventName: 'scale-input' }) scaleInput!: EventEmitter<KeyboardEvent>;
  /** Emitted when the value has changed. */
  @Event({ eventName: 'scale-change' })
  scaleChange!: EventEmitter<InputChangeEventDetail>;
  /** Emitted when the input has focus. */
  @Event({ eventName: 'scale-focus' }) scaleFocus!: EventEmitter<void>;
  /** Emitted when the input loses focus. */
  @Event({ eventName: 'scale-blur' }) scaleBlur!: EventEmitter<void>;
  /** Emitted when the input has focus. */
  @Event({ eventName: 'scale-focus-out' }) scaleFocusout!: EventEmitter<void>;
  /** Emitted on keydown. */
  @Event({ eventName: 'scale-keydown' })
  scaleKeyDown!: EventEmitter<KeyboardEvent>;
  /** Emitted on interactive icon click */
  @Event({ eventName: 'scale-interactive-icon-click' })
  scaleInteractiveIconClick!: EventEmitter<KeyboardEvent>;

  /** Whether the input element has focus */
  @State() hasFocus: boolean = false;

  /** "forceUpdate" hack, set it to trigger and re-render */
  @State() forceUpdate: string;

  private readonly internalId = generateUniqueId();

  componentWillLoad() {
    if (this.inputId == null) {
      this.inputId = 'input-search-' + this.internalId;
    }
  }

  componentDidRender() {
    // When `experimentalControlled` is true,
    // make sure the <input> is always in sync with the value.
    const value = this.value == null ? '' : this.value.toString();
    const input = this.hostElement.querySelector('input');
    if (this.experimentalControlled && input.value.toString() !== value) {
      input.value = value;
    }
  }

  handleInput = (event: Event) => {
    const target = event.target as HTMLInputElement | null;
    if (target) {
      this.value = target.value || '';
      this.emitChange();
    }
    if (this.experimentalControlled) {
      this.hostElement.querySelector('input').value = String(this.value);
      this.forceUpdate = String(Date.now());
    }
    emitEvent(this, 'scaleInput', event as KeyboardEvent);
  };

  handleChange = (event: Event) => {
    const target = event.target as HTMLInputElement | null;
    if (target) {
      this.value = target.value || '';
      this.emitChange();
    }
  };

  handleFocus = () => {
    this.hasFocus = true;
    emitEvent(this, 'scaleFocus');
  };

  handleFocusout = () => {
    this.hasFocus = false;
    emitEvent(this, 'scaleFocusout');
  };

  emitChange() {
    emitEvent(this, 'scaleChange', {
      value: this.value == null ? this.value : this.value.toString(),
    });
  }

  emitBlur = () => {
    emitEvent(this, 'scaleBlur');
  };

  emitKeyDown = (event: KeyboardEvent) => {
    emitEvent(this, 'scaleKeyDown', event);
  };

  emitInteractiveIconClick = (event: MouseEvent) => {
    emitEvent(this, 'scaleInteractiveIconClick', event);
  };

  getClearIconButton() {
    return (
      <scale-icon-button
        size="medium"
        part="clear-icon-button"
        onClick={() => (this.value = '')}
      >
        <scale-icon-action-close
          part="clear-icon"
          accessibility-title="close"
          size={24}
        />
      </scale-icon-button>
    );
  }

  render() {
    const ariaDetailedById = { 'aria-details': this.ariaDetailedId };
    const basePart = classNames(
      'base',
      this.hasFocus && 'focus',
      this.disabled && 'disabled'
    );
    return (
      <Host>
        {this.styles && <style>{this.styles}</style>}
        <div part={basePart}>
          <slot name="prefix-icon" />
          <input
            type="search"
            inputMode="search"
            part="input"
            placeholder={this.placeholder}
            value={this.value}
            {...(!!this.name ? { name: this.name } : {})}
            {...(!!this.inputAutofocus ? { autofocus: 'true' } : {})}
            required={this.required}
            minLength={this.minLength}
            maxLength={this.maxLength}
            id={this.inputId}
            onInput={this.handleInput}
            onChange={this.handleChange}
            onFocus={this.handleFocus}
            onFocusout={this.handleFocusout}
            onKeyDown={this.emitKeyDown}
            onBlur={this.emitBlur}
            disabled={this.disabled}
            autocomplete={this.inputAutocomplete}
            {...ariaDetailedById}
          ></input>
          {this.value ? (
            this.getClearIconButton()
          ) : (
            <div
              part="interactive-icon"
              onClick={this.emitInteractiveIconClick}
            >
              <slot name="suffix-icon" />
            </div>
          )}
        </div>
      </Host>
    );
  }
}
