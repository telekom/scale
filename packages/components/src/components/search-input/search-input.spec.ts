/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage } from '@stencil/core/testing';
import { SearchInput } from './search-input';

describe('Search', () => {
  let page;
  const TEST_VALUE = 'Test Value';
  beforeEach(async () => {
    page = await newSpecPage({
      components: [SearchInput],
      html: ` <scale-search-input>
                <scale-icon-action-search slot="prefix-icon"></scale-icon-action-search>
                <scale-icon-button size="medium" slot="suffix-icon">
                  <scale-icon-action-microphone
                    size="24"
                    accessibility-title="microphone" />
                </scale-icon-button>
              </scale-search-input>`,
    });
  });

  it('should match snapshot', async () => {
    expect(page.root).toMatchSnapshot();
  });

  it('should reflect attributes/props being set', async () => {
    const specPage = await newSpecPage({
      components: [SearchInput],
      html: `<scale-search-input
                name="search_name"
                invalid="true"
                max-length="2"
                min-length="1"
                placeholder="placeholder"
                disabled="true"
                required="true"
                value="value"
                input-id="input-search1"
                transparent="true"
                aria-detailed-id="id"
               >
                <scale-icon-action-search slot="prefix-icon"></scale-icon-action-search>
                <scale-icon-button size="medium" slot="suffix-icon">
                  <scale-icon-action-microphone
                    size="24"
                    accessibility-title="microphone" />
                </scale-icon-button
             ></scale-search-input>`,
    });

    expect(specPage.rootInstance.name).toBe('search_name');
    expect(specPage.rootInstance.invalid).toBe(true);
    expect(specPage.rootInstance.maxLength).toBe(2);
    expect(specPage.rootInstance.minLength).toBe(1);
    expect(specPage.rootInstance.placeholder).toBe('placeholder');
    expect(specPage.rootInstance.disabled).toBe(true);
    expect(specPage.rootInstance.required).toBe(true);
    expect(specPage.rootInstance.value).toBe('value');
    expect(specPage.rootInstance.inputId).toBe('input-search1');
    expect(specPage.rootInstance.ariaDetailedId).toBe('id');
  });

  it('should handle maxLength prop', async () => {
    page.root.maxLength = 15;
    await page.waitForChanges();
    expect(page.rootInstance.maxLength).toBe(15);
  });

  it('should handle minLength prop', async () => {
    page.root.minLength = 5;
    await page.waitForChanges();
    expect(page.rootInstance.minLength).toBe(5);
  });
});
