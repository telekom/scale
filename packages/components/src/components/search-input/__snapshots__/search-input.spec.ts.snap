// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should match snapshot 1`] = `
<scale-search-input>
  <mock:shadow-root>
    <div part="base">
      <slot name="prefix-icon"></slot>
      <input id="input-search-0" inputmode="search" name="Search" part="input" placeholder="Search" type="search" value="">
      <div part="interactive-icon">
        <slot name="suffix-icon"></slot>
      </div>
    </div>
  </mock:shadow-root>
  <scale-icon-action-search slot="prefix-icon"></scale-icon-action-search>
  <scale-icon-button size="medium" slot="suffix-icon">
    <scale-icon-action-microphone accessibility-title="microphone" size="24"></scale-icon-action-microphone>
  </scale-icon-button>
</scale-search-input>
`;
